File: 백준/Gold/20437. 문자열 게임 2/문자열 게임 2.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int T = Integer.parseInt(br.readLine());

        StringBuilder sb = new StringBuilder();
        
        while(T-- > 0) {
            String W = br.readLine().trim();
            int K = Integer.parseInt(br.readLine());

            Map<Character, List<Integer>> charIdx = new HashMap<>();

            for(int i = 0; i < W.length(); i++) {
                char c = W.charAt(i);

                charIdx.putIfAbsent(c, new ArrayList<>());
                charIdx.get(c).add(i);
            }

            int minLen = Integer.MAX_VALUE;
            int maxLen = -1;

            for(List<Integer> idxList : charIdx.values()) {
                if(idxList.size() < K) {
                    continue;
                }

                for(int i = 0; i <= idxList.size() - K; i++) {
                    int length = idxList.get(i + K - 1) - idxList.get(i) + 1;

                    minLen = Math.min(minLen, length);
                    maxLen = Math.max(maxLen, length);
                }
            }

            if(minLen == Integer.MAX_VALUE) {
                sb.append("-1\n");
            }
            else {
                sb.append(minLen).append(" ").append(maxLen).append("\n");
            }
        }
        System.out.println(sb.toString());
    }
}

File: 백준/Gold/20437. 문자열 게임 2/문자열 게임 2.py
import sys
input = sys.stdin.readline

T = int(input())

for _ in range(T):
    W = input().strip()
    K = int(input())
    char_idx = {}

    for i, char in enumerate(W):
        if char not in char_idx:
            char_idx[char] = []

        char_idx[char].append(i)

    min_len = float('inf')
    max_len = -1

    for idx_list in char_idx.values():
        if len(idx_list) < K:
            continue

        for i in range(len(idx_list) - K + 1):
            length = idx_list[i + K - 1] - idx_list[i] + 1

            min_len = min(min_len, length)
            max_len = max(max_len, length)

    if min_len == float('inf'):
        print(-1)
    else:
        print(min_len, max_len)